core/src/main/java/org/pentaho/di/core/database/DatabaseMetaInformation.java

```diff
@@ -0,0 +1,453 @@
+    return views;
+  }
+
+  /**
+   * @param views
+   *          The views to set.
+   */
+  public void setViews( String[] views ) {
+    this.views = views;
+  }
+
+  /**
+   * @param synonyms
+   *          The synonyms to set.
+   */
+  public void setSynonyms( String[] synonyms ) {
+    this.synonyms = synonyms;
+  }
+
+  /**
+   * @return Returns the synonyms.
+   */
+  public String[] getSynonyms() {
+    return synonyms;
+  }
+
+  /**
+   * @return Returns the procedures.
+   */
+  public String[] getProcedures() {
+    return procedures;
+  }
+
+  /**
+   * @param procedures
+   *          The procedures to set.
+   */
+  public void setProcedures( String[] procedures ) {
+    this.procedures = procedures;
+  }
+
+  public void getData( LoggingObjectInterface parentLoggingObject, ProgressMonitorListener monitor ) throws KettleDatabaseException {
+    if ( monitor != null ) {
+      monitor.beginTask( BaseMessages.getString( PKG, "DatabaseMeta.Info.GettingInfoFromDb" ), 8 );
+    }
+
+    Database db = new Database( parentLoggingObject, databaseMeta );
+
+    /*
+     * ResultSet tableResultSet = null;
+     *
+     * ResultSet schemaTablesResultSet = null; ResultSet schemaResultSet = null;
+     *
+     * ResultSet catalogResultSet = null; ResultSet catalogTablesResultSet = null;
+     */
+
+    try {
+      if ( monitor != null ) {
+        monitor.subTask( BaseMessages.getString( PKG, "DatabaseMeta.Info.ConnectingDb" ) );
+      }
+      db.connect();
+      if ( monitor != null ) {
+        monitor.worked( 1 );
+      }
+
+      if ( monitor != null && monitor.isCanceled() ) {
+        return;
+      }
+      if ( monitor != null ) {
+        monitor.subTask( BaseMessages.getString( PKG, "DatabaseMeta.Info.GettingMetaData" ) );
+      }
+      DatabaseMetaData dbmd = db.getDatabaseMetaData();
+      if ( monitor != null ) {
+        monitor.worked( 1 );
+      }
+
+      if ( monitor != null && monitor.isCanceled() ) {
+        return;
+      }
+      if ( monitor != null ) {
+        monitor.subTask( BaseMessages.getString( PKG, "DatabaseMeta.Info.GettingInfo" ) );
+      }
+      Map<String, String> connectionExtraOptions = databaseMeta.getExtraOptions();
+      if ( databaseMeta.supportsCatalogs() && dbmd.supportsCatalogsInTableDefinitions() ) {
+        ArrayList<Catalog> catalogList = new ArrayList<Catalog>();
+
+        String catalogFilterKey = databaseMeta.getPluginId() + "." + FILTER_CATALOG_LIST;
+        if ( ( connectionExtraOptions != null ) && connectionExtraOptions.containsKey( catalogFilterKey ) ) {
+          String catsFilterCommaList = connectionExtraOptions.get( catalogFilterKey );
+          String[] catsFilterArray = catsFilterCommaList.split( "," );
+          for ( int i = 0; i < catsFilterArray.length; i++ ) {
+            catalogList.add( new Catalog( catsFilterArray[i].trim() ) );
+          }
+        }
+        if ( catalogList.size() == 0 ) {
+          ResultSet catalogResultSet = dbmd.getCatalogs();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *catalogResultSet*. The resource is referenced at lines 234, 235. The resource is closed at line 241. There are other execution paths that do not contain closure statements, for example, when ResultSet.next throws an exception. Ensure *catalogResultSet* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).